packages/form-js-editor/src/features/dragging/Dragging.js

```diff
@@ -0,0 +1,350 @@
+import dragula from 'dragula';
+
+import { set as setCursor, unset as unsetCursor } from '../../render/util/Cursor';
+
+export const DRAG_CONTAINER_CLS = 'fjs-drag-container';
+export const DROP_CONTAINER_VERTICAL_CLS = 'fjs-drop-container-vertical';
+export const DROP_CONTAINER_HORIZONTAL_CLS = 'fjs-drop-container-horizontal';
+export const DRAG_MOVE_CLS = 'fjs-drag-move';
+export const DRAG_ROW_MOVE_CLS = 'fjs-drag-row-move';
+export const DRAG_COPY_CLS = 'fjs-drag-copy';
+export const DRAG_NO_DROP_CLS = 'fjs-no-drop';
+export const DRAG_NO_MOVE_CLS = 'fjs-no-move';
+export const ERROR_DROP_CLS = 'fjs-error-drop';
+
+/**
+ * @typedef { { id: String, components: Array<any> } } FormRow
+ */
+
+export default class Dragging {
+
+  /**
+   * @constructor
+   *
+   * @param { import('../../core/FormFieldRegistry').default } formFieldRegistry
+   * @param { import('../../core/FormLayouter').default } formLayouter
+   * @param { import('../../core/FormLayoutValidator').default } formLayoutValidator
+   * @param { import('../../core/EventBus').default } eventBus
+   * @param { import('../modeling/Modeling').default } modeling
+   */
+  constructor(formFieldRegistry, formLayouter, formLayoutValidator, eventBus, modeling) {
+    this._formFieldRegistry = formFieldRegistry;
+    this._formLayouter = formLayouter;
+    this._formLayoutValidator = formLayoutValidator;
+    this._eventBus = eventBus;
+    this._modeling = modeling;
+  }
+
+  /**
+   * Calculcates position in form schema given the dropped place.
+   *
+   * @param { FormRow } targetRow
+   * @param { any } targetFormField
+   * @param { HTMLElement } sibling
+   * @returns { number }
+   */
+  getTargetIndex(targetRow, targetFormField, sibling) {
+
+    /** @type HTMLElement */
+    const siblingFormFieldNode = sibling && sibling.querySelector('.fjs-element');
+    const siblingFormField = siblingFormFieldNode && this._formFieldRegistry.get(siblingFormFieldNode.dataset.id);
+
+    let targetIndex;
+
+    // (1) dropped before existing field => place before
+    if (siblingFormField) {
+      targetIndex = getFormFieldIndex(targetFormField, siblingFormField);
+    } else {
+
+      // (2) dropped in row => place at the end of row (after last field in row)
+      if (targetRow) {
+        targetIndex = getFormFieldIndex(
+          targetFormField,
+          this._formFieldRegistry.get(targetRow.components[targetRow.components.length - 1])
+        ) + 1;
+      } else {
+
+        // (3) dropped as last item
+        targetIndex = targetFormField.components.length;
+      }
+    }
+
+    return targetIndex;
```

Comment:
I think we can make it a little bit easier to read by using some early returns

```suggestion
if (siblingFormField) {
  return getFormFieldIndex(targetFormField, siblingFormField);
}

// (2) dropped in row => place at the end of row (after last field in row)
if (targetRow) {
  return (
    getFormFieldIndex(
      targetFormField,
      this._formFieldRegistry.get(
        targetRow.components[targetRow.components.length - 1],
      ),
    ) + 1
  );
}

// (3) dropped as last item
return targetFormField.components.length;
```